_source/_posts/2023-07-28-scim-workshop.md

```diff
@@ -0,0 +1,1144 @@
+
+```
+import { Router } from 'express';
+export const scimRoute = Router();
+```
+
+And you'll import the routes in `main.ts` like this:
+```
+import { scimRoute } from './scim';
+app.use('/scim/v2', scimRoute);
+```
+
+Note: `/scim/v2` will append to every SCIM route. You can change this static path as needed, as there isn't a specific URL path explicitly mentioned by the SCIM spec. However, [Okta's SCIM Docs](https://developer.okta.com/docs/guides/scim-provisioning-integration-prepare/main/#base-url) recommend using `/scim/v2` unless unusual factors in your environment necessitate a different URL. 
+
+## Build the SCIM interface:
+
+We'll need to build each CRUD endpoint required by the SCIM spec and format the responses in JSON. After that, we'll test our functions with Postman to see that they provide the responses required by the spec. 
+
+We'll need to use express and Prisma in `scim.ts`, so import them:
+
+```
+import express from 'express';
+import { Prisma, PrismaClient } from '@prisma/client';
+```
+
+Some SCIM endpoints in `scim.ts` will return user information from the Todo app's database. To easily retrieve this information from the SCIM endpoints, implement an `IUserSchema` interface and instantiate it as the defaultUserSchema. The information in the `IUserSchema` matches the SCIM spec, as you'll see soon.  
+
+```
+const prisma = new PrismaClient();
+
+// To funnel users into their designated orgs
+const ORG_ID = 1;
+interface IUserSchema {
+    schemas: string[];
+    userName?: string;
+    id?: string;
+    name?: {
+        givenName: string;
+        familyName: string;
+    };
+    emails?: { primary: boolean, value: string, type: string }[];
+    displayName?: string;
+    locale?: string;
+    meta?: {
+        resourceType: string;
+    }
+    externalId?: string;
+    groups?: [];
+    password?: string;
+    active?: boolean;
+    detail?: string;
+    status?: number;
+}
+
+const defaultUserSchema: IUserSchema = {
+    schemas: ['urn:ietf:params:scim:schemas:core:2.0:User'],
+    locale: 'en-US',
+    groups: [],
+    meta: {
+        resourceType: 'User'
+    }
+};
+```
+
+Finally, we are ready to get started! Let's work on our first CRUD endpoint. 
+
+### Create User Function - POST /Users - 
+
+Let's refer to the [SCIM spec](https://datatracker.ietf.org/doc/html/rfc7644#section-3.3) on creating a user. To create a user, the spec says that the client sends a POST request containing a "User" to the `/Users` endpoint. In response to the POST, the server signals a successful creation with an HTTP status code 201 (Created) and returns a representation of the user created. If the server determines that the creation of the requested user conflicts with existing users (e.g., a "User" resource with a duplicate "userName"), the server MUST return HTTP status code 409 (Conflict) with a "scimType" error code of "uniqueness," as per Section 3.12.
```

Comment:
This needs an explanation of who "the client" is. One of the most confusing parts about SCIM is that the "client" and "server" are kind of reversed compared to what you'd normally think of. Better to mention explicitly that the IdP is the client and will be making requests to your SaaS app. Again remember we're talking to SaaS app developers, so don't be afraid to mention that specifically, there's no need to try to make the language vague enough to apply to everyone.